source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+			category = DAMAGE;
+		if(category == COOL && effect == ENERGY)
+			category = ACTIVE_COOL;
+	}
+	this->category = category;
+	this->effect = effect;
+	this->secondary = secondary;
+}
+
+
+
+// Gets the data format name of the category, as used in the new syntax.
+string Attribute::GetCategoryName(const AttributeCategory category)
+{
+	if(category >= ATTRIBUTE_CATEGORY_COUNT || category < 0)
+		return "";
+	return categoryNames[category];
+}
+
+
+
+// Gets the data format name of the effect, as used in the new syntax. This also supports
+// multipliers, so for any effect E, passing E + ATTRIBUTE_EFFECT_COUNT will produce the name of the
+// multiplier effect.
+string Attribute::GetEffectName(const AttributeEffect effect)
+{
+	if(effect >= ATTRIBUTE_EFFECT_COUNT * 2)
+		return "relative " + GetEffectName(static_cast<AttributeEffect>(effect - 2 * ATTRIBUTE_EFFECT_COUNT));
+	if(effect >= ATTRIBUTE_EFFECT_COUNT)
+		return GetEffectName(static_cast<AttributeEffect>(effect - ATTRIBUTE_EFFECT_COUNT)) + " multiplier";
+	if(effect < 0)
+		return "";
+	return effectNames[effect];
+}
+
+
+
+// Calculates what the legacy name of the category-effect pair is. These results are cached
+// for faster access via GetLegacyName().
+string Attribute::CalculateLegacyName() const
+{
+	if(effect == -1)
+		return Attribute::GetCategoryName(category);
+	if(category == PASSIVE || category == -1)
+	{
+		if(effect == HULL)
+			return "hull";
+		if(effect == SHIELDS)
+			return "shields";
+		if(effect == RAMSCOOP)
+			return "ramscoop";
+		return Attribute::GetEffectName(effect) + (secondary == -1 ? "" : " " + Attribute::GetEffectName(secondary))
+				+ " capacity";
+	}
+
+	int effectType = effect % ATTRIBUTE_EFFECT_COUNT;
+	string effectName = Attribute::GetEffectName(static_cast<AttributeEffect>(effectType));
+	string secondaryName = Attribute::GetEffectName(static_cast<AttributeEffect>(secondary % ATTRIBUTE_EFFECT_COUNT));
+	string categoryName = Attribute::GetCategoryName(category);
+
+	if(effectType == HULL)
+		effectName = "hull";
+	if(category == PASSIVE)
+		categoryName = "";
+	else if((category == CLOAKING || category == THRUSTING || category == REVERSE_THRUSTING || category == TURNING)
+			&& (effect != -1 && static_cast<int>(category) != static_cast<int>(effect)))
+		categoryName += "ing";
+	else if(category == SHIELD_GENERATION && effectType != SHIELDS)
+		categoryName = "shield";
+	else if(category == HULL_REPAIR && effectType != HULL)
+		categoryName = "hull";
+	else if(category == HULL_REPAIR && (effect == HULL || effect == HULL + 2 * ATTRIBUTE_EFFECT_COUNT))
+		categoryName = "hull repair rate";
+	else if(category == HULL_REPAIR && effectType == HULL)
+		categoryName = "hull repair";
+	else if(category == AFTERBURNING)
+		categoryName = "afterburner";
+	else if(category == RESISTANCE && secondaryName != "")
+		categoryName += " " + secondaryName;
+	else if(category == DAMAGE && effectType == PIERCING)
+		categoryName = "";
```

Comment:
why's this empty and also doesnt that mean it should be looked at earlier with the PASSIVE one?

Reply:
Piercing is a damage-related attribute, but it's called `piercing`, not `piercing damage`, unlike `shield damage` or `slowing damage`. It's one of the many edge cases in the old system.

Reply:
Also, in that function, I was trying to make the code more logical for how specific attribute names are calculated, that's why there are similar conditions repeated in some places. I think it's easier to tweak than something minimized.